lab4/lab4.rst

```diff
@@ -0,0 +1,230 @@
+
+-  Build a controller for a set-associative and non-blocking cache.
+
+-  Analyze the performance of different cache designs with a simulator.
+
+Description
+===========
+
+In this lab, you will build a cache model of a set-associative non-blocking caches in C++.
+Computer architects often use high-level languages like C++ to *model* hardware in early research and development of new designs.
+Using high-level languages gives architects more freedom and flexibility to quickly iterate with designs than using hardware description languages or actually making hardware.
+
+You will be given a simulation framework and an example direct-mapped cache implementation.
+You will build your new cache models in this framework and evaluate them.
+You will be graded in two ways:
+#. Correctly implementing your cache models (e.g., the correct number of tag bits).
+#. Understanding the tradeoffs between different designs.
+
+**DO NOT MODIFY THE GIVEN CODE IN ANY WAY. DOING SO WILL LEAD TO AN AUTOMATIC ZERO.**
+(Unless expressly stated below.)
+
+**EXTRA CREDIT**: This is my first time giving this assignment.
+I have written this code over the last couple of days, so it's likely there are bugs.
+If you're unsure if you've found a bug, post on Piazza.
+If you find a bug, please open a pull request on Github_.
+**You will receive 5 points of extra credit for each pull request I merge**.
+
+.. _Github: https://github.com/jlpteaching/ECS154B/
+
+Details
+=======
+
+Given Code
+----------
+
+You will find the code in the ``lab4/provided/`` directory.
+The code is heavily documented.
+Please see the code for all details.
+
+There is a Makefile included that will build the ``cache_simulator``.
+The cache simulator currently has a very simple test implemented.
+In ``main.cc``, a processor, a cache, and a memory are instantiated.
+Then, the simulation is executed.
+
+Processor
+~~~~~~~~~
+
+The processor drives the simulation.
+It sends memory requests to the cache by calling ``receiveRequest`` on the cache object.
+The processor will continually send requests to the cache until the cache responds that it is blocked (``receiveRequest`` returns ``false``).
+There requests may be sent multiple in a single *tick* or cycle.
+
+The test that is run is initialized in ``Processor::createRecords()``.
+This function populates a queue which is the "trace" of memory accesses.
+See ``processor.cc`` for details.
+
+Feel free to add your own tests and/or modify the tests in ``createRecords``.
+When we test your code, we will use a different implementation of ``Processor::createRecords``.
+
+Cache
+~~~~~
+
+You will be inheriting from the ``Cache`` object.
+See the code in ``cache.hh`` for details of the interface you will implement.
+You will need to implement the following functions:
+- ``receiveRequest``: Called by the Processor when sending a new request.
+- ``receiveMemResponse``: Called by memory when it has finished reading the data previously requested by the cache.
```

Comment:
Are there equivalent functions for the cache to call to send data to the Processor and requests to the memory? It might make sense to have a place where we list all critical functions in one place.

Reply:
Good point. I'll update with this information.